Enhance report overview

tclaus opened the pull request:
This PR enhances the admin/moderators report overview by 
 - adding a view that shows recent reports and the action done (Item Deleted or not) 
 - adding a statistics view (Who reports and who is the originator of a reported item) 
 - a once 'reviewed report can be deleted retrospectively
 - let podsmins show the originator even /after/ a reported item was deleted
 - links to the originator profile directly (and not to the Search view) 
 
**Heads up!**
It needs a migration! 
Two fields are added: a 'action' textfield that is used to describe what happened to the report. Its free text, but filled based on the action. 
The second fields holds the diaspora_handel for the originator. This is needed because after deleting a report this information is gone. 

**What is the outcome?**
A Podmin now has more insights in old reports and in most reporter and most originators. A valuable base for more decisions what to do. 

**Future**
List of reports grows without limits - a paginated view would be nice, but that was currently above my skill level.


Screenshots

 (Added date time)
<img width="910" alt="Bildschirmfoto 2021-04-28 um 08 11 07" src="https://user-images.githubusercontent.com/501326/116355491-3652f100-a7fa-11eb-88b1-9bea3bd9eb44.png">

( deleted or reviewed reports) 
<img width="898" alt="Bildschirmfoto 2021-04-28 um 08 10 56" src="https://user-images.githubusercontent.com/501326/116355311-f55adc80-a7f9-11eb-96d8-0c73e0c21640.png">

(statistics with links to the authors / originators) 
<img width="939" alt="Bildschirmfoto 2021-04-28 um 08 11 17" src="https://user-images.githubusercontent.com/501326/116355560-4e2a7500-a7fa-11eb-8be9-31a3f9a730c2.png">


Comments, Ideas, critics... welcome!

tclaus committed:
Adds a detailed report overview

With undecided / reviewed and statistics page
Already reviewed reports stay visible
Enhances Spam/user control

fixes #8239

tclaus commented:
This Feature was requested in this discussion: 
https://discourse.diasporafoundation.org/t/better-abilities-for-podmins-for-spam-analytics-and-controls/3896/6
and on this Issue: #8222 

Solving my own issues, nice isn't it?

diaspora-code-review commented on app/helpers/report_helper.rb:
```diff
@@ -8,17 +8,36 @@ module ReportHelper
     case (item = report.item)
     when Post
-      raw t("report.post_label", content: link_to(post_message(item), post_path(item.id)))
+      raw t("report.post_label", content: link_to(truncated_post_message(item), post_path(item.id)))
```
Rails/OutputSafety: Tagging a string as html safe may be a security risk.

Flaburgan commented:
Hey! This is very needed, thank you for working on this. I'm going to try it and do a review, but two quick remarks:
- I guess the reports only appear in the "Reviewed" tab if they have been reviewed, meaning they have been either "marked as reviewed" or "deleted". That means a boolean "deleted" or if you prefer an enum ["ignored", "deleted"] should be enough to be stored in the database, instead of a free text field. Strings displayed in the interface shouldn't be stored in the DB, it makes them untranslatable, and is bad practice anyway
- Originator isn't clear to me. Why not using "Author" as everywhere else?

tclaus commented:
The Images in the first comment are not matching the text. 

> * Originator isn't clear to me. Why not using "Author" as everywhere else?

"Reporter" and "Author" then - I missed this. 


Do you mean to use the Database' ENUM type? Like the Idea.

Flaburgan commented:
>Do you mean to use the Database' ENUM type? Like the Idea.

I mean your mockup gave the impression that "No action" was the text stored in the database. This must be avoided, the state (deleted or ignored) as to be stored in the DB, then the UI will display a label depending of the state.

Flaburgan reviewed (changes requested):
This is **GREAT**, I love it!

I did some remarks but that's mostly in good shape already. But you would need to also add tests. I started to add them in https://github.com/diaspora/diaspora/pull/8035 but I'm actually not able to navigate to /report in my PR. If you help me fix that PR you should then be able to rebase on me and complete the tests there.

diaspora-code-review commented on app/assets/javascripts/app/views.js:
```diff
@@ -103,28 +103,27 @@ app.views.Base = Backbone.View.extend({
-      }
+    if (evt) { evt.preventDefault(); }
+    var report = {
+      item_id: this.model.id,
```
Identifier 'item_id' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/views.js:
```diff
@@ -103,28 +103,27 @@ app.views.Base = Backbone.View.extend({
+    if (evt) { evt.preventDefault(); }
+    var report = {
+      item_id: this.model.id,
+      item_type: $(evt.currentTarget).data("type"),
```
Identifier 'item_type' is not in camel case.

diaspora-code-review commented on db/migrate/20210421064041_add_action_to_report.rb:
```diff
@@ -0,0 +1,9 @@
+  def change
+    add_column :reports, :action, :string
+
+    Report.update_all(action: Report::STATUS_NO_ACTION)
```
Rails/SkipsModelValidations: Avoid using `update_all` because it skips validations.

diaspora-code-review commented on features/step_definitions/report_steps.rb:
```diff
@@ -0,0 +1,52 @@
+    find("tr", text: "Deleted")
+  end
+end
+
```
Layout/TrailingEmptyLines: 1 trailing blank lines detected.

tclaus commented:
@Flaburgan I rebased this on Enhance Report Form

goobertron commented on app/models/report.rb:
```diff
@@ -4,32 +4,41 @@ class Report < ApplicationRecord
 
-  after_commit :send_report_notification, :on => :create
+  STATUS_DELETED = "deleted"
+  STATUS_NO_ACTION = "no action"
```
'no action' seems fine to me in that context. It makes clear that a decision has been made to take no action on that content. The label 'ignored' is potentially confusing: it could mean that the person who looked at the report decided to ignore the report (didn't make a decision), or that the content was 'ignored' i.e. hidden, or something else.

I'm not sure why some of the reports that have 'no action' next to them also have eye and bin icons to the right of them and others don't – is that clear to anyone using this feature?

tclaus replied:
As Fla found out, with introduction of the new "action" field, no existing Reports have a default initial value. I've added a migration and set all existing Reports to 'NO_ACTION', so every reviewed report should now have the action buttons.

tclaus commented:
@Flaburgan , @SuperTux88 
Besides the (old and already existing) issues in Linter and Rubocop, is there anything which blocks a review?

Flaburgan commented:
https://github.com/diaspora/diaspora/pull/8035 is adding tests on that part of the code so we said we should merge it first, but it's blocked by some weird Jasmine flickering tests. If you want to help there you're welcome!

SuperTux88 commented on app/controllers/report_controller.rb:
```diff
@@ -9,14 +9,17 @@ class ReportController < ApplicationController
   def index
-    @reports = Report.where(reviewed: false)
+    @unreviewed_reports = Report.join_originator.where(reviewed: false).order(created_at: :desc)
+    @reviewed_reports = Report.join_originator.where(reviewed: true).order(created_at: :desc)
```
I don't know if this is a good idea. Big pods can have a lot of reports, and loading all of them might be a performance problem, especially since this is always loaded, even if you only want to view new unreviewed reports.

So this either needs some sort of pagination, or at least only load the reviewed reports if you actually want to load that tab.

tclaus replied:
Limiting for the last 100 for now.

SuperTux88 commented on app/helpers/report_helper.rb:
```diff
@@ -5,19 +5,40 @@
+    case (item = report.item)
+    when Post
+      link_to("", post_path(item.id),
+              {title:  "View reported element",
```
This text should probably be translated.

Also, either write "View reported post" here, or if you keep "element", it could used for both which I think is fine as well (so less duplication needed, and only the actual links needs to be different).

SuperTux88 commented on app/helpers/notifier_helper.rb:
```diff
@@ -11,6 +11,15 @@ def post_message(post, opts={})
     rendered.presence || post_page_title(post)
   end
 
+  def truncated_post_message(post)
+    if post.respond_to? :message
+      plain_text = post.message.try(:plain_text_without_markdown).presence || post_page_title(post)
+      truncate(plain_text, length: 300)
+    else
+      I18n.t "notifier.a_post_you_shared"
+    end
+  end
+
```
This has nothing to do with the notifier, so it shouldn't be added here.

Also, as commented in https://github.com/diaspora/diaspora/pull/8035#discussion_r956791512, I would just use `post_page_title(post)` in all cases (which already is truncated), so this isn't even needed at all (also I think this should already be fixed in #8035, so you wouldn't need to do anything).

tclaus replied:
Moved (and changed) to report_helper

SuperTux88 commented on app/helpers/notifier_helper.rb:
```diff
@@ -21,4 +30,13 @@ def comment_message(comment, opts={})
     end
   end
+
+  def truncated_comment_message(comment)
+    if comment.post.public?
+      plain_text = comment.message.plain_text_without_markdown
+      truncate(plain_text, length: 180)
+    else
+      I18n.t "notifier.a_limited_post_comment"
+    end
+  end
```
Same as above, this isn't needed, and #8035 should already use `comment.message.title` directly, so you don't need to do anything here.

tclaus replied:
Removed as suggested

SuperTux88 commented on app/models/report.rb:
```diff
@@ -4,21 +4,31 @@ class Report < ApplicationRecord
   belongs_to :post, optional: true
   belongs_to :comment, optional: true
   belongs_to :item, polymorphic: true
+  delegate :author, to: :item
```
Adding the author of the item as `author` here is very confusing, as `report.author` then looks like this is the author if the report, please add at least a `prefix: true` here.

Also, as the item might get deleted, it would probably be smarter to instead use the new column which stores the author of the reported item even if the item itself gets deleted.

tclaus replied:
Hm, if the item is deleted the author is also not every interesting.  I'll take the prefix.

tclaus replied:
Hm 2: There exists already a 'reported_author' in the report model. This is already a better name for what it means. I drop the delegate and use the existing code.

SuperTux88 commented on app/views/report/_pending.haml:
```diff
@@ -0,0 +1,48 @@
+    .panel.panel-default
+      - username = report.user.username
+      .panel-heading
+        .reporter.pull-right
+          != t("report.reported_label", person: link_to(username, user_profile_path(username)))
+        .author
+          %span.reason-label
+            != "#{t('report.author')}:"
+          %span
+            = report.author.name
+        .created-at
+          %span.reason-label
+            != "#{t('report.created_at')}:"
+          %span
+            = I18n.l(report.created_at, format: :short)
+        .reason
+          %span.reason-label
+            = t("report.reason_label")
+          %span
+            = report.text
```
I think this can be very confusing, what is about what. We have "Created at" which is about the report, and then we have "Author" which is the author of the reported item. I don't really know what's the best solution for this, maybe changing the text to "Reported Author/Person" makes it clearer, or maybe having two columns, one for the stuff that is about the report, and one that is about the reported item? But really, I'm mostly throwing ideas here and don't know what is best.

Also, as this information here is also interesting even if the item was deleted (for example by the author of the reported item), I would restructure the code there to always show the same things. The only difference is that if the item was deleted you can't show the text/link of the reported item anymore.

tclaus replied:
`report.author `is now prefixed with item_author. I also changed the localized text. But you are right, we have the report author, message and creation date as well as the items (post, comments) attributes of the same meaning.

tclaus replied:
Ah, I now see, there is already a `reported_author` in the model. I use this.

SuperTux88 commented on app/views/report/_checked.haml:
```diff
@@ -0,0 +1,40 @@
+      %td
+        = report.item_type
+      %td
+        - unless report.originator_diaspora_handle.nil?
+          = link_to(report.originator_diaspora_handle, "/people/#{report.originator_guid}")
```
If we don't know the author of the reported item anymore (because it was deleted before the migration), then I would show "Unknown" here instead of just blank.

SuperTux88 reviewed (changes requested):
As this is depending on #8035, this isn't a full review yet (I maybe have missed some things between all the changes of the other PR), and you might also want to wait until #8035 is done before working on some of my comments (as there will be conflicts). But there is also already some feedback you can work on without having conflicts, so I already wanted to submit that before the other PR is done.

Also, I couldn't add a comment to that file, as it doesn't have a single line, but you accidentally deleted `tmp/pids/.gitkeep`.

tclaus commented on db/migrate/20210421064041_add_action_to_report.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddActionToReport < ActiveRecord::Migration[5.2]
+  def change
+    add_column :reports, :action, :string
```
I have merged the two migrations together

tclaus commented on app/views/report/_pending.haml:
```diff
@@ -0,0 +1,43 @@
+            = report.text
+      .panel-body
+        .content
+          = report_content(report)
```
Added in line 12 report.author.name is a delegate to report.item.author, aka the author of the reported post.

tclaus commented on db/migrate/20210424054604_add_user_fields_to_reports.rb:
```diff
@@ -0,0 +1,19 @@
+
+class AddUserFieldsToReports < ActiveRecord::Migration[5.2]
+  def change
+    add_column :reports, :originator_diaspora_handle, :string, index: true
```
With ID you mean the numeric id column or the unique guid?

SuperTux88 replied:
with "person ID" I mean the numeric `id` column of the `people` table.

tclaus commented on db/migrate/20210421064041_add_action_to_report.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    add_column :reports, :action, :string
+
+    # rubocop:disable Rails/SkipsModelValidations
```
Thanks for this, I did not know... - and removed this migration at the end.

tclaus commented on features/step_definitions/report_steps.rb:
```diff
@@ -0,0 +1,51 @@
+  find_link("Reviewed").click
+end
+
+Then(/^I should see the reviewed report with decision No Action$/) do
+  within("#checked") do
+    find("tr", text: "No Action")
+  end
+end
+
+Then(/^I should see the reviewed report with decision Deleted$/) do
+  within("#checked") do
+    find("tr", text: "Deleted")
+  end
+end
```
Thanks for the hint.

tclaus commented on features/desktop/report.feature:
```diff
@@ -0,0 +1,87 @@
+    Then I should see a report by "alice@alice.alice" with reason "That's my reason" on comment "Bob comment"
+    And "alice@alice.alice" should have received an email with subject "A new comment was marked as offensive"
+
+    Scenario: A checked report appears in the Reviewed section as reviewed
+      Given I sign in as "alice@alice.alice"
+      And I am on the public stream page
+      When I hover over the ".stream-element"
+      And I click to report the post
+      Then I should see the report modal
+      When I fill in "report-reason-field" with "That's my reason"
+      And I submit the form
+      When I go to the report page
+      And I mark report as reviewed
+      When I open the reviewed tab on the report page
+      Then I should see the reviewed report with decision No Action
```
Shifted to the left

diaspora-code-review commented on app/assets/javascripts/app/app.js:
```diff
@@ -155,6 +156,29 @@ var app = {
+      $("#reportModal").modal("hide");
+      let textarea = document.getElementById("report-reason-field");
+      let report = {
+        item_id: form.dataset.reportId,
```
Identifier 'item_id' is not in camel case.

diaspora-code-review commented on app/assets/javascripts/app/app.js:
```diff
@@ -155,6 +156,29 @@ var app = {
+      let textarea = document.getElementById("report-reason-field");
+      let report = {
+        item_id: form.dataset.reportId,
+        item_type: form.dataset.reportType,
```
Identifier 'item_type' is not in camel case.

diaspora-code-review commented on app/models/report.rb:
```diff
@@ -4,33 +4,43 @@ class Report < ApplicationRecord
 
   def reported_author
-    item.author if item
+    return Person.find(reported_author_id) if reported_author_id.present?
```
Layout/EmptyLineAfterGuardClause: Add empty line after guard clause.

diaspora-code-review commented on app/models/report.rb:
```diff
@@ -4,33 +4,43 @@ class Report < ApplicationRecord
-    return unless Report.where(item_id: item_id, item_type: item_type).exists?(user_id: user_id)
-
-    errors.add(:base, "You cannot report the same post twice.")
+    if Report.where(item_id: item_id, item_type: item_type).exists?(user_id: user_id)
```
Style/GuardClause: Use a guard clause (`return unless Report.where(item_id: item_id, item_type: item_type).exists?(user_id: user_id)`) instead of wrapping the code inside a conditional expression.

diaspora-code-review commented on app/models/report.rb:
```diff
@@ -4,33 +4,43 @@ class Report < ApplicationRecord
-    return unless Post.find_by(id: item_id).nil? && Comment.find_by(id: item_id).nil?
-
-    errors.add(:base, "Post or comment was already deleted or doesn't exists.")
+    if Post.find_by(id: item_id).nil? && Comment.find_by(id: item_id).nil?
```
Style/GuardClause: Use a guard clause (`return unless Post.find_by(id: item_id).nil? && Comment.find_by(id: item_id).nil?`) instead of wrapping the code inside a conditional expression.

diaspora-code-review commented on db/migrate/20221030224657_add_action_fields_to_reports.rb:
```diff
@@ -0,0 +1,20 @@
+class AddActionFieldsToReports < ActiveRecord::Migration[6.1]
```
Style/FrozenStringLiteralComment: Missing frozen string literal comment.

diaspora-code-review commented on db/migrate/20221030224657_add_action_fields_to_reports.rb:
```diff
@@ -0,0 +1,20 @@
+class AddActionFieldsToReports < ActiveRecord::Migration[6.1]
+
```
Layout/EmptyLinesAroundClassBody: Extra empty line detected at class body beginning.

diaspora-code-review commented on db/migrate/20221030224657_add_action_fields_to_reports.rb:
```diff
@@ -0,0 +1,20 @@
+
+    Report.find_each do |report|
+      # get reported author id from item before item gets deleted
+      if report.reported_author.present?
```
Style/IfUnlessModifier: Favor modifier `if` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.

diaspora-code-review commented on db/migrate/20221030224657_add_action_fields_to_reports.rb:
```diff
@@ -0,0 +1,20 @@
+      if report.reported_author.present?
+        report.reported_author_id = report.reported_author.id
+      end
+      if report.item.present?
```
Style/IfUnlessModifier: Favor modifier `if` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.

diaspora-code-review commented on db/migrate/20221030224657_add_action_fields_to_reports.rb:
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+class AddActionFieldsToReports < ActiveRecord::Migration[6.1]
```
Layout/EmptyLineAfterMagicComment: Add an empty line after magic comments.

tclaus commented:
@Flaburgan  I rebased this to your #8035  and fixed most of my issues here.

Flaburgan commented:
So because the report modal got merged, I rebased this MR, squashed the commits and fixed the front-end pronto problems. There are still a lot of them but I'd like help to resolve those.
I did not even run the code yet, I will obviously try this and polish the UI as we said in the other PR.

Flaburgan commented:
@SuperTux88 about the remark you made about the design now that multi lines reports are possible.

It looks fine (the design isn't broken) but we are losing the line break:

![image](https://github.com/diaspora/diaspora/assets/930064/57efda4e-d0c7-4305-9c73-6fc01ea9e404)
![image](https://github.com/diaspora/diaspora/assets/930064/1d48e28b-3391-4dbc-92e3-995f698fb967)

Is there anything else you want me to check on the front-end side?

SuperTux88 commented on app/models/report.rb:
```diff
@@ -50,12 +61,16 @@ def destroy_reported_item
+
+  def mark_as_reviewed(with_action=STATUS_NO_ACTION)
+    Report.where(item_id: item_id, item_type: item_type)
+          .update_all(reviewed: true, action: with_action)
```
There is no `action` column anymore, did you accidentally drop the migrations?

tclaus replied:
Eagle-eye

SuperTux88 commented on app/helpers/notifier_helper.rb:
```diff
@@ -4,21 +4,23 @@ module NotifierHelper
   def post_message(post, opts={})
-    rendered = opts[:html] ? post.message&.markdownified_for_mail : post.message&.plain_text_without_markdown
-    rendered.presence || post_page_title(post)
+    if post.respond_to? :message
+      post.message.try(:plain_text_without_markdown).presence || post_page_title(post)
+    else
+      I18n.t "notifier.a_post_you_shared"
+    end
   end
 
   # @param comment [Comment] The comment to process.
-  # @param opts [Hash] Optional hash.  Accepts :html parameter.
   # @return [String] The formatted comment.
   def comment_message(comment, opts={})
     if comment.post.public?
-      opts[:html] ? comment.message.markdownified_for_mail : comment.message.plain_text_without_markdown
+      comment.message.plain_text_without_markdown
     else
-      I18n.translate "notifier.a_limited_post_comment"
+      I18n.t "notifier.a_limited_post_comment"
     end
   end
```
These helpers also get used for the notification mails (hence the name "NotifierHelper"), you can't just completely change it's functionality and drop html support. But it looks like you aren't even using these helpers anymore, so maybe just revert these changes?

tclaus replied:
Reverted this.

SuperTux88 reviewed (changes requested):
This isn't a full review, I only had a quick look. But it looks like something went wrong when rebasing this, as at least the migration was dropped (which probably also explains why the tests failed).

> It looks fine (the design isn't broken) but we are losing the line break

Well, the reports contain plain newlines, you would need to replace them with html-newlines (but ensure that no other html tags can be injected, so not just mark the string as html-safe)

tclaus committed:
Adds a detailed report overview

With undecided / reviewed and statistics page
Already reviewed reports stay visible
Enhances Spam/user control
fixing style issues

fixes #8239

tclaus commented on app/controllers/report_controller.rb:
```diff
@@ -38,7 +43,23 @@ def create
+  end
+
+  def statistics_by_reporter
+    sql = "select count(*), diaspora_handle, guid from reports
```
Have rewritten this

tclaus commented on app/controllers/report_controller.rb:
```diff
@@ -38,7 +43,23 @@ def create
+  end
+
+  def statistics_by_author
+    sql = "select count(*), originator_diaspora_handle, guid from reports
```
Have rewritten this as ActiveRecord

tclaus committed:
fixing error..

tclaus committed:
Fixing Review issues

adding migration
solving style errors

tclaus committed:
Back to add_column in migration

The 'add_reference' did not work as expected.

tclaus committed:
Back to add_column in migration

The 'add_reference' did not work as expected.

tclaus committed:
Fixing a test - in tests messages my come in in almost no time, so order by created_at might fail the test.

tclaus committed:
Fixing a test - in tests messages my come in in almost no time, so order by created_at might fail the test.

Flaburgan commented:
I am facing more and more spam, this feature would be very welcomed. It would be even better if the [diaspora_spam.rb](https://gist.github.com/jhass/9104029) script by @jhass could be integrated. Currently tests are failing but I don't understand why.
